### PR TITLE
stats: json optimization of histograms

### LINKED ISSUE #3102: Optimize JSON Model for Histograms
As part of supporting JSON model for histograms, currently supported Quantiles are being repeated for every histogram with two arrays interval and cumulative. An optimized approach could be to just send supported Quantiles once and repeat the interval and cumulative array for each histogram.

### PR DESCRIPTION
**Pull request by ramaraochavali**
Signed-off-by: Rama <rama.rao@salesforce.com>
Description:
optimizes the json representation of histograms by sending all supported_quantile once and computed_quantlie array multiple times.

Risk Level: Low
Testing: Manual
Docs Changes: NA
Release Notes: NA

Fixes #3102
Fixes https://github.com/envoyproxy/envoy/issues/3210

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@mattklein123 @jmarantz @mrice32 
I screwed up the original PR https://github.com/envoyproxy/envoy/pull/3184 trying to correct DCO. Sorry. 
Here is the new PR.
Addressed all the comments in the previous PR and added docs.
PTAL.

File: docs/root/operations/admin.rst [hunk 1]
```diff
@@ -190,7 +190,48 @@ The fields are:
...
+            "name": "example.counter",
+            "value": 0
+        },
+        {
+            "name": "example.gauge",
+            "value": 0
+        },
+        {
+            "histograms": {
+                "supported_quantiles": [...],
```
Comment (mrice32) [inline docs/root/operations/admin.rst]: Could you add real values here, so the user knows what the array will look like (as in, is it an array of `doubles` or subobjects)?

File: docs/root/operations/admin.rst [hunk 2]
```diff
@@ -190,7 +190,48 @@ The fields are:
...
+                "computed_quantiles": [
+                    {
+                        "name": "example.histogram",
+                        "values": [
+                            {
+                                "interval": null,
```
Comment (mrice32) [inline docs/root/operations/admin.rst]: As above, it would be nice to see an example of a histogram with non-null quantiles values since that will be the common case.  But it'd also be nice to keep some sort of `null` example so the user knows what a no-data hist will look like.  

Side note: will this ever print histograms that have no samples in both cumulative and internal since that would mean its `used()` method would return `false`?

File: docs/root/operations/admin.rst (hunk 1 above)
Comment (ramaraochavali) [reply docs/root/operations/admin.rst]: will add.

File: docs/root/operations/admin.rst (hunk 2 above)
Comment (ramaraochavali) [reply docs/root/operations/admin.rst]: I will make those changes. and regarding used, it currently prints all histograms. And I think it is a good point, it would be good filter the unused ones. I will make that change as well.

**Comment by ramaraochavali** [comment]
release test is failing with server_fuzz_test unrelated to this.

**Comment by mattklein123** [comment]
@ramaraochavali try merging master, @htuch put in a fix that might help.

File: source/common/stats/thread_local_store.cc [hunk 3]
```diff
@@ -96,8 +96,8 @@ void ThreadLocalStoreImpl::shutdownThreading() {
 }
 
 void ThreadLocalStoreImpl::mergeHistograms(PostMergeCb merge_complete_cb) {
-  ASSERT(!merge_in_progress_);
   if (!shutting_down_) {
+    ASSERT(!merge_in_progress_);
```
Comment (ramaraochavali) [inline source/common/stats/thread_local_store.cc]: i just  moved this check because, I hit the same issue in CI. Since https://github.com/envoyproxy/envoy/pull/3289 fixes it properly we should be good.

File: docs/root/operations/admin.rst [hunk 4]
```diff
@@ -190,9 +190,118 @@ The fields are:
...
-  .. http:get:: /stats?format=prometheus
+  If a histogram is not updated during an interval, the ouput will have null for all the quantiles.
+  
+  Example histogram output:
+
+.. code-block:: json
+
+  {
+    "histograms": {
+    "supported_quantiles": [
```
Comment (jmarantz) [inline docs/root/operations/admin.rst]: the indentation in this description looks wrong.  This line should be indented one more level than the previous one.  Do you think it also makes sense, doc purpose, to put these 9 quantiles each on one line?

File: docs/root/operations/admin.rst [hunk 5]
```diff
@@ -190,9 +190,118 @@ The fields are:
...
+      99,
+      99.9,
+      100
+    ],
+    "computed_quantiles": [
+      {
+        "name": "cluster.external_auth_cluster.upstream_cx_length_ms",
+        "values": [
+          {
+            "interval": 0,
```
Comment (jmarantz) [inline docs/root/operations/admin.rst]: for doc legibility purposes, consider this format:
```
        "values": [
           {"interval": 0, "cumulative": 0},
           {"interval": 0, "cumulative": 0},
           {"interval": 1.0435787, "cumulative": 1.0435787},
           ....
```
I assume the whitespace output is not actually what the json serializer actually emits, right?  It would omit all whitespace?

File: docs/root/operations/admin.rst [hunk 6]
```diff
@@ -190,9 +190,118 @@ The fields are:
...
+            "cumulative": 229.57333
+          },
+          {
+            "interval": null,
+            "cumulative": 260
+          }
+        ]
+      }
+    ]
+   }
```
Comment (jmarantz) [inline docs/root/operations/admin.rst]: I think the missing indentation earlier forces a few lines of 1-char indent.  All this is hard to see because of all the newlines above.

File: test/integration/integration_admin_test.cc [hunk 7]
```diff
@@ -161,10 +161,37 @@ TEST_P(IntegrationAdminTest, Admin) {
...
   Json::ObjectSharedPtr statsjson = Json::Factory::loadFromString(response->body());
   EXPECT_TRUE(statsjson->hasObject("stats"));
   EXPECT_STREQ("application/json", ContentType(response));
+  EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+
+  uint64_t histogram_count = 0;
+  for (Json::ObjectSharedPtr obj_ptr : statsjson->getObjectArray("stats")) {
+    if (obj_ptr->hasObject("histograms")) {
+      histogram_count++;
+      const Json::ObjectSharedPtr& histograms_ptr = obj_ptr->getObject("histograms");
```
Comment (jmarantz) [inline test/integration/integration_admin_test.cc]: I'm confused here: you are looking for 'histograms' plural so I was expecting  another loop over all the histograms.  Can you explain?

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: There is a toplevel  "histograms" element that has supported_quantiles array and computed_quantiels, which is again an array of actual histogram computed values in the form of "name" and "values" (interval and cumulative) - the last array is repeated as many times as there are histograms"

Comment (jmarantz) [reply test/integration/integration_admin_test.cc]: I'm still not entirely clear on the cardinality of variouis things from looking at the code.  If we actually add a second histogram and make the test expect 2?

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: no. it is still 1. The computed_quantiles array size would be equivalent to the number of histograms.

File: test/integration/integration_admin_test.cc [hunk 8]
```diff
@@ -161,10 +161,37 @@ TEST_P(IntegrationAdminTest, Admin) {
...
+      // Validate that both supported_quantiles and computed_quantiles are present in JSON.
+      EXPECT_TRUE(histograms_ptr->hasObject("supported_quantiles"));
+      EXPECT_TRUE(histograms_ptr->hasObject("computed_quantiles"));
+
+      const std::vector<Json::ObjectSharedPtr>& computed_quantiles =
+          histograms_ptr->getObjectArray("computed_quantiles");
+      EXPECT_GT(computed_quantiles.size(), 0);
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: this count would be equivalent to number of histograms. I do not know exactly how many will come that is why I have added this GT zero condition.

File: docs/root/operations/admin.rst [hunk 9]
```diff
@@ -203,102 +203,40 @@ The fields are:
...
-        ]
-      }
-    ]
-   }
+      "supported_quantiles": [
+        0,25,50,75,90,95,99,99.9,100
+      ],
+      "computed_quantiles": [
+       {
+          "name": "cluster.external_auth_cluster.upstream_cx_length_ms",
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply docs/root/operations/admin.rst]: It is two only. I checked by enabling white space. Github is showing like that.

File: docs/root/operations/admin.rst [hunk 10]
```diff
@@ -190,9 +190,56 @@ The fields are:
...
+.. code-block:: json
+
+  {
+    "histograms": {
+      "supported_quantiles": [
+        0, 25, 50, 75, 90, 95, 99, 99.9, 100
+      ],
+      "computed_quantiles": [
+       {
+          "name": "cluster.external_auth_cluster.upstream_cx_length_ms",
```
Comment (jmarantz) [inline docs/root/operations/admin.rst]: You have this:
```
+       {
+          "name": "cluster.external_auth_cluster.upstream_cx_length_ms",
```
I think it should be this:
```
+       {
+         "name": "cluster.external_auth_cluster.upstream_cx_length_ms",
```

File: source/common/stats/thread_local_store.cc (hunk 3 above)
Comment (jmarantz) [reply source/common/stats/thread_local_store.cc]: It'd be great to get that PR in first to avoid affecting this one :)

File: test/integration/integration_admin_test.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/integration/integration_admin_test.cc]: Isn't the number of histograms predictable in this test?

File: test/integration/integration_admin_test.cc (hunk 7 above)
Comment (jmarantz) [reply test/integration/integration_admin_test.cc]: Can you please add a second histogram to this test, so computed_quantiles array has length 2?

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: I am not adding any histograms in the test. I just grab whatever the output is from /stats?format=json and validate that it has histograms, it would definitely some (but non predictable). I am not setting up any thing for this test.

Comment (jmarantz) [reply test/integration/integration_admin_test.cc]: Ah, very cool.  Let me rephrase my request.  Can you make a new test where you add some specific histogram data with more than one histogram, and check the JSON output?

**Comment by ramaraochavali** [comment]
@jmarantz since that is a private method used only at this place, I thought this test is sufficient. I am not sure if we can add tests for private methods in the current test infrastructure. Will have to check. If you think it adds value, I will do it next week

**Comment by jmarantz** [comment]
you can structure the test exactly as it is here, if you like, via the admin interface, if that's easier; I just think it's reasonable to control the data.  It's also possible to test private methods by friending the test-class and putting a helper there, or adding a TestingPeer class which is friended, if that helps unit-tests.

**Comment by ramaraochavali** [comment]
@jmarantz  i will try adding the test next week. I did not get how I can structure the test with admin interface though. How will clear the existing histograms/add the histograms I want in the admin test - i do not have access to stats_store I guess. I have not looked at the code. Are you suggesting some thing different? May not getting your suggestion...

**Comment by jmarantz** [comment]
RE test for >1 histogram: I would actually prefer you add it as a unit test anyway, which sounds easier.  The 'private' issue is easily dealt with by either friending the test-class or a TestingPeer class, either of which can act as in intermediary to access a private method.

File: test/server/http/admin_test.cc [hunk 11]
```diff
@@ -23,12 +24,89 @@
...
 using testing::_;
 
 namespace Envoy {
 namespace Server {
 
+/**
+ * This is a heap test allocator that works similar to how the shared memory allocator works in
+ * terms of reference counting, etc.
+ */
+class TestAllocator : public Stats::RawStatDataAllocator {
```
Comment (ramaraochavali) [inline test/server/http/admin_test.cc]: @jmarantz added a unit test that verifies the json format. PTAL. I had to copy this TestAllocator class as is from thread_local_store_test.cc as I found that it is the easiest way to generate histograms with stores. I do not know if there is any way to make this class across both the files, with out changing much..

File: docs/root/operations/admin.rst [hunk 12]
```diff
@@ -190,9 +190,56 @@ The fields are:
...
+          "values": [
+            {"interval": null, "cumulative": 0},
+            {"interval": null, "cumulative": 0},
+            {"interval": null, "cumulative": 1.0435787},
+            {"interval": null, "cumulative": 1.0941565},
+            {"interval": null, "cumulative": 2.0860023},
+            {"interval": null, "cumulative": 3.0665233},
+            {"interval": null, "cumulative": 6.046609},
+            {"interval": null, "cumulative": 229.57333},
+            {"interval": null, "cumulative": 260}
```
Comment (jmarantz) [inline docs/root/operations/admin.rst]: This looks great now. Much easier to read.  Thanks!

File: test/server/http/admin_test.cc [hunk 13]
```diff
@@ -49,6 +127,146 @@ class AdminFilterTest : public testing::TestWithParam<Network::Address::IpVersio
...
+
+  store_->mergeHistograms([]() -> void {});
+
+  EXPECT_CALL(*this, free(_));
+
+  std::map<std::string, uint64_t> all_stats;
+
+  std::string actual_json = statsAsJsonHandler(all_stats, store_->histograms());
+
+  const std::string expected_json = R"EOF({
```
Comment (jmarantz) [inline test/server/http/admin_test.cc]: This reminds me; do we always generate pretty-printed JSON?  I think it would be better to deliver it compact in production, though it's nice to have the test be pretty-printed for test.cc legibility.

File: test/server/http/admin_test.cc (hunk 11 above)
Comment (jmarantz) [reply test/server/http/admin_test.cc]: Can you move the TestAllocator class to test/test_common/utility.h so you don't have duplicated code?

Comment (ramaraochavali) [reply test/server/http/admin_test.cc]: Yes. Moved

File: test/server/http/admin_test.cc (hunk 13 above)
Comment (ramaraochavali) [reply test/server/http/admin_test.cc]: Yes. Currently we do generate pretty-printed JSON. Now I have added a parameter to control that and defaulted it to false and changed the test to print in pretty form. PTAL.

File: test/test_common/utility.h [hunk 14]
```diff
@@ -340,8 +342,51 @@ class TestHeaderMapImpl : public HeaderMapImpl {
...
+
+    return stat_ref.get();
+  }
+
+  void free(RawStatData& data) override {
+    if (--data.ref_count_ > 0) {
+      return;
+    }
+
+    for (auto i = stats_.begin(); i != stats_.end(); i++) {
```
Comment (jmarantz) [inline test/test_common/utility.h]: This was pre-existing, but you have a map here, can't we just remove it via the name?  I don't really care about perf for the test but I think the code would just be:
```
  if (stats_.erase(std::string(data.name()) == 0) {
    FAIL();
  }
```
which is fewer lines :)

File: source/server/http/admin.cc [hunk 15]
```diff
@@ -518,43 +520,65 @@ AdminImpl::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
...
+      for (size_t i = 0; i < histogram->intervalStatistics().supportedQuantiles().size(); ++i) {
+        Value quantile_obj;
+        quantile_obj.SetObject();
+        Value interval_value;
+        if (!std::isnan(histogram->intervalStatistics().computedQuantiles()[i])) {
+          interval_value.SetDouble(histogram->intervalStatistics().computedQuantiles()[i]);
+        }
+        quantile_obj.AddMember("interval", interval_value, allocator);
+        Value cumulative_value;
+        if (!std::isnan(histogram->cumulativeStatistics().computedQuantiles()[i])) {
```
Comment (jmarantz) [inline source/server/http/admin.cc]: Probably worth commenting here that as we are positionally associating  this entry with a quantile, so we need to put in the {null, null} entry to keep other data aligned.

**Comment by ramaraochavali** [comment]
@jmarantz addressed the two comments. Should be good now.
@mattklein123 I think we are good here. Can you PTAL ?
